src/pymor/algorithms/simplify.py

```diff
@@ -97,6 +118,75 @@ def action_LincombOperator(self, op):
             op = op.with_(operators=ops, coefficients=coeffs)
         return op
 
+    @match_class(AdjointOperator)
+    def action_AdjointOperator(self, op):
+        inner = self.apply(op.operator)
+
+        # If the inside is a LincombOperator, distribute adjoint over the sum:
+        # (∑ c_i A_i)^* = ∑ conj(c_i) A_i^*
+        if isinstance(inner, LincombOperator):
+            ops = [AdjointOperator(o, source_product=op.source_product, range_product=op.range_product)
+                   for o in inner.operators]
+            coeffs = [c.conjugate() for c in inner.coefficients]
+            return LincombOperator(ops, coeffs)
+
+        # If the inside is a ConcatenationOperator, distribute adjoint over the concatenation:
+        # (A @ B @ C)^* = C^* @ B^* @ A^*
+        if isinstance(inner, ConcatenationOperator):
+            if len(inner.operators) == 1:
+                return AdjointOperator(inner.operators[0], source_product=op.source_product,
+                                       range_product=op.range_product)
+            else:
+                last, *middle, first = inner.operators[::-1]
+                adj_factors = ([AdjointOperator(last, source_product=op.source_product)] +
+                            [op.H for op in middle] +
+                            [AdjointOperator(first, range_product=op.range_product)])
+
+                return ConcatenationOperator(adj_factors)
+
+        return AdjointOperator(inner, source_product=op.source_product, range_product=op.range_product)
+
+    @match_class(BlockOperator)
+    def action_BlockOperator(self, op):
+        # recursively expand all children
+        op = self.replace_children(op)
+        nrows, ncols = op.blocks.shape
+
+        # Build the constant part: copy non-Lincomb blocks; Lincomb blocks start as zero
+        const_blocks = np.full((nrows, ncols), None, dtype=object)
+        for (i, j), b in np.ndenumerate(op.blocks):
+            if not isinstance(b, LincombOperator):
+                const_blocks[i, j] = b
+
+        expanded_terms = []
+        for (i, j), b in np.ndenumerate(op.blocks):
+            if not isinstance(b, LincombOperator):
+                continue
+            for c_k, a_k in zip(b.coefficients, b.operators, strict=True):
+                if isinstance(c_k, ParameterFunctional):
+                    blocks_k = np.full((nrows, ncols), None, dtype=object)
+                    blocks_k[i, j] = a_k
+                    expanded_terms.append((c_k, BlockOperator(blocks_k, range_spaces=op.range.subspaces,
+                                                            source_spaces=op.source.subspaces)))
+                else:
+                    cur = const_blocks[i, j]
+                    if cur is None:
+                        const_blocks[i, j] = c_k * a_k
+                    else:
+                        const_blocks[i, j] += c_k * a_k
```

Comment:
```suggestion
                blocks_k = np.full((nrows, ncols), None, dtype=object)
                blocks_k[i, j] = a_k
                expanded_terms.append((c_k, BlockOperator(blocks_k, range_spaces=op.range.subspaces,
                                                        source_spaces=op.source.subspaces)))
```

Maybe I'm confused, but isn't this more in line with that we have discussed? The else clause would cause `const_part` to have `LincombOperators` as blocks, but the main purpose of `expand` is to move `LincombOperators` from the inside to the outside? When `contract` is called afterwards, both version should give the same result.